File: init_users.js
#!/usr/bin/env node
const fs = require('fs');
const bcrypt = require('bcryptjs');

function readMembers(filename) {
    const content = fs.readFileSync(filename, 'utf-8');
    return content.split('\n').filter(line => line.trim());
}

function createManittoChain(names) {
    const shuffled = [...names];
    // Fisher-Yates shuffle
    for (let i = shuffled.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
    }

    // Create circular chain
    const chain = [];
    for (let i = 0; i < shuffled.length; i++) {
        const fromPerson = shuffled[i];
        const toPerson = shuffled[(i + 1) % shuffled.length];
        chain.push({ from: fromPerson, to: toPerson });
    }

    return chain;
}

async function generatePasswordHash(password) {
    return await bcrypt.hash(password, 10);
}

async function generateSQL(membersFile, outputFile) {
    // Read members
    const names = readMembers(membersFile);
    console.log(`Found ${names.length} members: ${names.join(', ')}`);

    // Create manitto chain
    const manittoChain = createManittoChain(names);
    console.log('\nManitto chain:');
    manittoChain.forEach(({ from, to }) => {
        console.log(`  ${from} -> ${to}`);
    });

    // Generate SQL
    const sqlLines = [];

    // Header
    sqlLines.push('-- Initialize users table with manitto assignments');
    sqlLines.push('-- Generated automatically by init_users.js');
    sqlLines.push('');

    // Clear existing data
    sqlLines.push('-- Clear existing data');
    sqlLines.push('DELETE FROM public.messages;');
    sqlLines.push('DELETE FROM public.push_tokens;');
    sqlLines.push('DELETE FROM public.users;');
    sqlLines.push('');

    // Insert users
    sqlLines.push('-- Insert users');
    const userData = {};

    for (const name of names) {
        const email = `${name}@kakaomobility.com`;
        const initialPassword = name;
        const passwordHash = await generatePasswordHash(initialPassword);

        sqlLines.push(
            `INSERT INTO public.users (name, email, initial_password, password_hash) ` +
            `VALUES ('${name}', '${email}', '${initialPassword}', '${passwordHash}');`
        );

        userData[name] = {
            email,
            initialPassword
        };
    }

    sqlLines.push('');

    // Update manitto relationships
    sqlLines.push('-- Update manitto relationships');
    manittoChain.forEach(({ from, to }) => {
        sqlLines.push(
            `UPDATE public.users ` +
            `SET manitto_to = (SELECT id FROM public.users WHERE name = '${to}') ` +
            `WHERE name = '${from}';`
        );
    });

    sqlLines.push('');

    // Set manitto_from (reverse relationship)
    manittoChain.forEach(({ from, to }) => {
        sqlLines.push(
            `UPDATE public.users ` +
            `SET manitto_from = (SELECT id FROM public.users WHERE name = '${from}') ` +
            `WHERE name = '${to}';`
        );
    });

    sqlLines.push('');
    sqlLines.push('-- Verify results');
    sqlLines.push('SELECT ');
    sqlLines.push('  u.name,');
    sqlLines.push('  u.email,');
    sqlLines.push('  u.initial_password,');
    sqlLines.push('  u_from.name as manitto_from,');
    sqlLines.push('  u_to.name as manitto_to');
    sqlLines.push('FROM public.users u');
    sqlLines.push('LEFT JOIN public.users u_from ON u.manitto_from = u_from.id');
    sqlLines.push('LEFT JOIN public.users u_to ON u.manitto_to = u_to.id');
    sqlLines.push('ORDER BY u.name;');

    // Write to file
    fs.writeFileSync(outputFile, sqlLines.join('\n'));

    console.log(`\nSQL script generated: ${outputFile}`);
    console.log('\nUser credentials:');
    names.forEach(name => {
        console.log(`  ${name}: ${userData[name].email} / ${userData[name].initialPassword}`);
    });
}

// Run
generateSQL('members.txt', 'init-users.sql')
    .then(() => console.log('\nDone!'))
    .catch(err => {
        console.error('Error:', err);
        process.exit(1);
    });

File: init_users.py
#!/usr/bin/env python3
import random
import hashlib

def generate_password_hash(password: str) -> str:
    """Generate password hash using SHA-256"""
    return hashlib.sha256(password.encode('utf-8')).hexdigest()

def read_members(filename: str) -> list[str]:
    """Read member names from file"""
    with open(filename, 'r') as f:
        return [line.strip() for line in f if line.strip()]

def create_manitto_chain(names: list[str]) -> list[tuple[str, str]]:
    """Create a circular chain of manitto relationships"""
    shuffled = names.copy()
    random.shuffle(shuffled)

    # Create circular chain: shuffled[0] -> shuffled[1] -> ... -> shuffled[n-1] -> shuffled[0]
    chain = []
    for i in range(len(shuffled)):
        from_person = shuffled[i]
        to_person = shuffled[(i + 1) % len(shuffled)]
        chain.append((from_person, to_person))

    return chain

def generate_sql(members_file: str, output_file: str):
    """Generate SQL script to initialize database"""

    # Read members
    names = read_members(members_file)
    print(f"Found {len(names)} members: {', '.join(names)}")

    # Create manitto chain
    manitto_chain = create_manitto_chain(names)
    print("\nManitto chain:")
    for from_person, to_person in manitto_chain:
        print(f"  {from_person} -> {to_person}")

    # Generate SQL
    sql_lines = []

    # Header
    sql_lines.append("-- Initialize users table with manitto assignments")
    sql_lines.append("-- Generated automatically by init_users.py")
    sql_lines.append("")

    # Clear existing data
    sql_lines.append("-- Clear existing data")
    sql_lines.append("DELETE FROM public.messages;")
    sql_lines.append("DELETE FROM public.push_tokens;")
    sql_lines.append("DELETE FROM public.users;")
    sql_lines.append("")

    # Insert users
    sql_lines.append("-- Insert users")
    user_data = {}
    for name in names:
        email = f"{name}@kakaomobility.com"
        initial_password = name
        password_hash = generate_password_hash(initial_password)

        sql_lines.append(
            f"INSERT INTO public.users (name, email, initial_password, password_hash) "
            f"VALUES ('{name}', '{email}', '{initial_password}', '{password_hash}');"
        )
        user_data[name] = {
            'email': email,
            'initial_password': initial_password
        }

    sql_lines.append("")

    # Update manitto relationships
    sql_lines.append("-- Update manitto relationships")
    for from_person, to_person in manitto_chain:
        sql_lines.append(
            f"UPDATE public.users "
            f"SET manitto_to = (SELECT id FROM public.users WHERE name = '{to_person}') "
            f"WHERE name = '{from_person}';"
        )

    sql_lines.append("")

    # Set manitto_from (reverse relationship)
    for from_person, to_person in manitto_chain:
        sql_lines.append(
            f"UPDATE public.users "
            f"SET manitto_from = (SELECT id FROM public.users WHERE name = '{from_person}') "
            f"WHERE name = '{to_person}';"
        )

    sql_lines.append("")
    sql_lines.append("-- Verify results")
    sql_lines.append("SELECT ")
    sql_lines.append("  u.name,")
    sql_lines.append("  u.email,")
    sql_lines.append("  u.initial_password,")
    sql_lines.append("  u_from.name as manitto_from,")
    sql_lines.append("  u_to.name as manitto_to")
    sql_lines.append("FROM public.users u")
    sql_lines.append("LEFT JOIN public.users u_from ON u.manitto_from = u_from.id")
    sql_lines.append("LEFT JOIN public.users u_to ON u.manitto_to = u_to.id")
    sql_lines.append("ORDER BY u.name;")

    # Write to file
    with open(output_file, 'w') as f:
        f.write('\n'.join(sql_lines))

    print(f"\nSQL script generated: {output_file}")
    print("\nUser credentials:")
    for name in names:
        print(f"  {name}: {user_data[name]['email']} / {user_data[name]['initial_password']}")

if __name__ == "__main__":
    generate_sql('members.txt', 'init-users.sql')
